contracts/sol6/fprConversionRates/QtyStepConversionRates.sol

```diff
@@ -0,0 +1,429 @@
+    /// @dev If user want to use this for more than 14 token, they should use setBaseRate
+    function setBaseRateWithEmptyCompactData(
+        IERC20Ext[] calldata tokens,
+        uint256[] calldata baseBuy,
+        uint256[] calldata baseSell,
+        uint256 blockNumber
+    ) external onlyOperator {
+        uint256 compactDataArrayIndex = uint256(-1);
+        for (uint256 ind = 0; ind < tokens.length; ind++) {
+            IERC20Ext token = tokens[ind];
+            TokenData memory data = tokenData[token];
+            require(data.listed, "unlisted token");
+            if (compactDataArrayIndex == uint256(-1)) {
+                compactDataArrayIndex = data.compactDataArrayIndex;
+            }
+            require(
+                compactDataArrayIndex == data.compactDataArrayIndex,
+                "CompactData from different slot"
+            );
+        }
+        // fill in dummy data to compact data
+        bytes14[] memory buy = new bytes14[](1);
+        buy[0] = bytes14(0);
+        bytes14[] memory sell = new bytes14[](1);
+        sell[0] = bytes14(0);
+        uint256[] memory indices = new uint256[](1);
+        indices[0] = compactDataArrayIndex;
+        setBaseRate(tokens, baseBuy, baseSell, buy, sell, blockNumber, indices);
+    }
+
+    function setQtyStepFunction(
+        IERC20Ext token,
+        int256[] calldata xBuy,
+        int256[] calldata yBuy,
+        int256[] calldata xSell,
+        int256[] calldata ySell
+    ) external onlyOperator {
+        require(xBuy.length == yBuy.length, "xBuy-yBuy length mismatch");
+        require(xSell.length == ySell.length, "xSell-ySell length mismatch");
+        require(xBuy.length <= MAX_STEPS_IN_FUNCTION, "xBuy too big");
+        require(xSell.length <= MAX_STEPS_IN_FUNCTION, "xSell too big");
+        require(tokenData[token].listed, "unlisted token");
+
+        tokenBuyQtySteps[token].length = xBuy.length;
+        for (uint256 i = 0; i < xBuy.length; i++) {
+            tokenBuyQtySteps[token].data[i] = StepFunction(
+                safeInt128(xBuy[i]),
+                safeInt128(yBuy[i])
+            );
+        }
+        tokenSellQtySteps[token].length = xSell.length;
+        for (uint256 i = 0; i < xSell.length; i++) {
+            tokenSellQtySteps[token].data[i] = StepFunction(
+                safeInt128(xSell[i]),
+                safeInt128(ySell[i])
+            );
+        }
+    }
+
+    function setValidRateDurationInBlocks(uint256 duration) external onlyAdmin {
+        validRateDurationInBlocks = duration;
+    }
+
+    function enableTokenTrade(IERC20Ext token) external onlyAdmin {
+        require(tokenData[token].listed, "unlisted token");
+        require(
+            tokenControlInfo[token].minimalRecordResolution != 0,
+            "tokenControlInfo is required"
+        );
+        tokenData[token].enabled = true;
+    }
+
+    function disableTokenTrade(IERC20Ext token) external onlyOperator {
```

Comment:
More suitable to use `alerter` here, but `operator` also fine

Reply:
Please change this to `onlyAlerter`, so it is consistent with our current conversion rate implementation, otherwise our reserve team may need to update the code